### PR TITLE
Minimal implementation of Decorators

### PR DESCRIPTION
**Pull request by rbuckton**
Minimal implementation of Decorators

### CODE REVIEW

**Comment by paulvanbrenk** [comment]
Can you add testcases for sourcemaps

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -100,11 +100,18 @@ module ts {
         let globalIterableType: ObjectType;
 
         let anyArrayType: Type;
+        let globalTypedPropertyDescriptorType: ObjectType;
+        let globalClassDecoratorType: ObjectType;
+        let globalClassAnnotationType: ObjectType;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: let's try to limit the number of concepts users have to deal with and keep the decorator only falvor.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -376,7 +384,7 @@ module ts {
                     //   }
                     //
                     case SyntaxKind.ComputedPropertyName:
-                        let grandparent = location.parent.parent;
+                        grandparent = location.parent.parent;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: not used.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -408,6 +416,26 @@ module ts {
                             break loop;
                         }
                         break;
+                    case SyntaxKind.Decorator:
+                        if (location.parent) {
+                            lastLocation = location;
+                            grandparent = location.parent.parent;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: grandparent for all should be sufficient

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -8093,6 +8121,10 @@ module ts {
             checkFunctionLikeDeclaration(node);
         }
 
+        function checkIncompleteDeclaration(node: Declaration) {
+            error(node, Diagnostics.Declaration_expected);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: should be in the parser.

File: src/compiler/diagnosticInformationMap.generated.ts [hunk 5]
```diff
@@ -162,6 +162,11 @@ module ts {
...
         Export_assignment_cannot_be_used_when_targeting_ECMAScript_6_or_higher_Consider_using_export_default_instead: { code: 1203, category: DiagnosticCategory.Error, key: "Export assignment cannot be used when targeting ECMAScript 6 or higher. Consider using 'export default' instead." },
         Cannot_compile_external_modules_into_amd_or_commonjs_when_targeting_es6_or_higher: { code: 1204, category: DiagnosticCategory.Error, key: "Cannot compile external modules into amd or commonjs when targeting es6 or higher." },
+        Decorators_are_only_supported_on_class_members_when_targeting_ECMAScript_5_or_higher: { code: 1205, category: DiagnosticCategory.Error, key: "Decorators are only supported on class members when targeting ECMAScript 5 or higher." },
+        Decorators_are_not_valid_on_this_declaration_type: { code: 1206, category: DiagnosticCategory.Error, key: "Decorators are not valid on this declaration type." },
+        Argument_to_ambient_decorator_must_be_constant_expression: { code: 1207, category: DiagnosticCategory.Error, key: "Argument to ambient decorator must be constant expression." },
```
Comment (mhegazy) [inline src/compiler/diagnosticInformationMap.generated.ts]: not needed.

File: src/compiler/emitter.ts [hunk 6]
```diff
@@ -5000,6 +5126,172 @@ module ts {
...
+                    return;
+                }
+
+                if (languageVersion >= ScriptTarget.ES5) {
+                    emitStart(member);
+                    emitDecorateStart(decorators);
+                    emitTargetOfClassElement(node, member);
+                    write(", ");
+                    emitExpressionForPropertyName(name);
+                    write(");");
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: nit, can we call this emitDecoratorEnd();

**Comment by rbuckton** [comment]
Merged manually into master

File: src/compiler/parser.ts [hunk 7]
```diff
@@ -4968,14 +5113,15 @@ module ts {
 
         function parseDeclaration(): ModuleElement {
             let fullStart = getNodePos();
+            let decorators = parseDecorators();
```
Comment (jayphelps) [inline src/compiler/parser.ts]: This allows you to use decorators *before* the `export` keyword.
```javascript
@decorator export default class Foo {}
```
But if [this](https://github.com/wycats/javascript-decorators) is the spec we're following, this isn't actually included in the grammar. Only:

```javascript
export default @decorator class Foo {}
```

Comment (jayphelps) [inline src/compiler/parser.ts]: Subscribers, discussion is being had here https://github.com/wycats/javascript-decorators/issues/7#issuecomment-86247666

**Pull request closed without merging**